src/libraries/Common/tests/System/Net/Sockets/SocketTestExtensions.cs

```diff
@@ -49,5 +51,32 @@ public static (Socket, Socket) CreateConnectedSocketPair()
 
             return (client, server);
         }
+
+        // Tries to connect within the provided timeout interval
+        // Useful to speed up "can not connect" assertions on Windows
+        public static bool TryConnect(this Socket socket, EndPoint remoteEndpoint, int millisecondsTimeout)
+        {
+            using var mre = new ManualResetEvent(false);
+            using var sea = new SocketAsyncEventArgs()
+            {
+                RemoteEndPoint = remoteEndpoint,
+                UserToken = mre
+            };
+
+            sea.Completed += (s, e) =>
+            {
+                if (e.UserToken is ManualResetEvent mreInner) mreInner.Set();
+            };
+
+            bool pending = socket.ConnectAsync(sea);
+            if (!pending || mre.WaitOne(millisecondsTimeout))
+            {
+                return sea.SocketError == SocketError.Success;
+            }
+
+            sea.UserToken = null; // Letting the event handler know that we timed out
```

Comment:
Not really necessary; you could have the event handler always set it and not null it out. Then you can remove this like and simplify the handler to just:
```C#
sea.Completed += (s, e) => ((ManualResetEventSlim)e.UserToken).Set();
```

Reply:
I'm doing this to address the race condition you pointed out in your previous comment. Without that a pending operation may try to set this handle.

Wit the current logic, I think we are safe unless an unexpected exception is thrown.

Reply:
Ok you are right, a race condition is still present, but it seems unlikely to cause trouble in test code.

Is it indeed safer to just leave the MRE undisposed instead?

Reply:
> it seems unlikely to cause trouble in test code.

I can almost guarantee it would happen at some point in CI and cause a spurious test failure.

> Is it indeed safer to just leave the MRE undisposed instead?

Yes

Reply:
Ok, fixed.